File: development/NET/samples/server/UpdateMode/CS/MyTransaction.cs
using System;
using Softing.OPCToolbox;
using Softing.OPCToolbox.Server;

namespace UpdateMode
{
	public class MyTransaction : DaTransaction
	{
		#region Public Methods
		//--------------------

		public MyTransaction(
			EnumTransactionType aTransactionType,
			DaRequest[] requestList,
			uint aSessionKey)
			:
			base(
			aTransactionType,
			requestList,
			aSessionKey)
		{
		}   //	end constructor

		public override int HandleReadRequests()
		{
			int count = Requests.Count;

			for (int i = 0; i < count; i++)
			{
				DaRequest request = (DaRequest)Requests[i];
				if ((request != null) && (request.ProgressRequestState(EnumRequestState.CREATED, EnumRequestState.INPROGRESS) == true))
				{
					MyDaAddressSpaceElement element = request.AddressSpaceElement as MyDaAddressSpaceElement;

					if (element == null)
					{
						request.Result = EnumResultCode.E_FAIL;
					}
					else
					{
						if (request.PropertyId == 0)
						{
							if (element.Type == MyDaAddressSpaceElement.TYPE_POLL_SEC)
							{
								// get address space element value
								DateTime now = DateTime.Now;
								ValueQT cacheValue = new ValueQT(now.Second, EnumQuality.GOOD, now);
								request.Value = cacheValue;
								request.Result = EnumResultCode.S_OK;
							}
							else if (element.Type == MyDaAddressSpaceElement.TYPE_POLL_MIN)
							{
								// get address space element value
								DateTime now = DateTime.Now;
								ValueQT cacheValue = new ValueQT(now.Minute, EnumQuality.GOOD, now);
								request.Value = cacheValue;
								request.Result = EnumResultCode.S_OK;
							}
							else if (element.Type == MyDaAddressSpaceElement.TYPE_REPORT_MIN)
							{
								//	this is a direct device read request
								DateTime now = DateTime.Now;
								ValueQT cacheValue = new ValueQT(now.Minute, EnumQuality.GOOD, now);
								request.Value = cacheValue;
								request.Result = EnumResultCode.S_OK;
							}
							else if (element.Type == MyDaAddressSpaceElement.TYPE_REPORT_SEC)
							{
								//	this is a direct device read request
								DateTime now = DateTime.Now;
								ValueQT cacheValue = new ValueQT(now.Second, EnumQuality.GOOD, now);
								request.Value = cacheValue;
								request.Result = EnumResultCode.S_OK;
							}   //	end if ... else
						}
						else
						{
							// get property value
							// get the value from the address space element
							element.GetPropertyValue(request);
						}   //	end if ... else
					}   //	end if ... else
				}   //	end if
			}   //	end for
			return CompleteRequests();
		}   //	end HandleReadRequests


		//--
		#endregion
	}   //	end class MyTransaction
}   //	end namespace

File: development/NET/samples/server/UpdateMode/VB/MyTransaction.vb
Imports System
Imports Softing.OPCToolbox
Imports Softing.OPCToolbox.Server

Namespace UpdateMode

	Public Class MyTransaction
		Inherits DaTransaction

#Region "Public Methods"
		'--------------------

		Public Sub New(ByVal aTransactionType As EnumTransactionType, ByVal requestList As DaRequest(), ByVal aSessionKey As UInt32)
			MyBase.New(aTransactionType, requestList, aSessionKey)
		End Sub

		Public Overloads Overrides Function HandleReadRequests() As Integer
			Dim count As Integer = Requests.Count

			Dim i As Integer = 0
			While i < count
				Dim request As DaRequest = DirectCast(Requests(i), DaRequest)
				If request.ProgressRequestState(EnumRequestState.CREATED, EnumRequestState.INPROGRESS) Then
					Dim element As MyDaAddressSpaceElement = CType(request.AddressSpaceElement, MyDaAddressSpaceElement)

					If element Is Nothing Then
						request.Result = EnumResultCode.E_FAIL
					Else
						If request.PropertyId = 0 Then

							If element.Type = MyDaAddressSpaceElement.TYPE_POLL_SEC Then

								' get address space element value
								Dim now As DateTime = DateTime.Now
								Dim cacheValue As New ValueQT(now.Second, [Enum].ToObject(GetType(EnumQuality), EnumQuality.GOOD), now)
								request.Value = cacheValue
								request.Result = EnumResultCode.S_OK
							ElseIf element.Type = MyDaAddressSpaceElement.TYPE_POLL_MIN Then

								' get address space element value
								Dim now As DateTime = DateTime.Now
								Dim cacheValue As New ValueQT(now.Minute, [Enum].ToObject(GetType(EnumQuality), EnumQuality.GOOD), now)
								request.Value = cacheValue
								request.Result = EnumResultCode.S_OK
							ElseIf element.Type = MyDaAddressSpaceElement.TYPE_REPORT_SEC Then

								' get address space element value
								'    this is a direct device read request
								Dim now As DateTime = DateTime.Now
								Dim cacheValue As New ValueQT(now.Second, [Enum].ToObject(GetType(EnumQuality), EnumQuality.GOOD), now)
								request.Value = cacheValue
								request.Result = EnumResultCode.S_OK
							ElseIf element.Type = MyDaAddressSpaceElement.TYPE_REPORT_MIN Then

								' get address space element value
								'    this is a direct device read request
								Dim now As DateTime = DateTime.Now
								Dim cacheValue As New ValueQT(now.Minute, [Enum].ToObject(GetType(EnumQuality), EnumQuality.GOOD), now)
								request.Value = cacheValue
								request.Result = EnumResultCode.S_OK
							End If
						Else
							' get property value
							' get the value from the address space element
							element.GetPropertyValue(request)
						End If
					End If
				End If
				i = i + 1
			End While
			Return CompleteRequests()
		End Function


		'--
#End Region
	End Class
End Namespace

